core/arch/arm/kernel/ldelf_loader.c

```diff
@@ -0,0 +1,320 @@
+ * Copyright (c) 2020, Arm Limited
+ */
+
+#include <assert.h>
+#include <kernel/ldelf_loader.h>
+#include <ldelf.h>
+#include <mm/mobj.h>
+#include <mm/vm.h>
+
+extern uint8_t ldelf_data[];
+extern const unsigned int ldelf_code_size;
+extern const unsigned int ldelf_data_size;
+extern const unsigned int ldelf_entry;
+
+/* ldelf has the same architecture/register width as the kernel */
+#ifdef ARM32
+static const bool is_arm32 = true;
+#else
+static const bool is_arm32;
+#endif
+
+static TEE_Result alloc_and_map_ldelf_fobj(struct user_mode_ctx *uctx,
+					   size_t sz, uint32_t prot,
+					   vaddr_t *va)
+{
+	size_t num_pgs = ROUNDUP(sz, SMALL_PAGE_SIZE) / SMALL_PAGE_SIZE;
+	struct fobj *fobj = fobj_ta_mem_alloc(num_pgs);
+	struct mobj *mobj = mobj_with_fobj_alloc(fobj, NULL);
+	TEE_Result res = TEE_SUCCESS;
+
+	fobj_put(fobj);
+	if (!mobj)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = vm_map(uctx, va, num_pgs * SMALL_PAGE_SIZE,
+		     prot, VM_FLAG_LDELF, mobj, 0);
+	mobj_put(mobj);
+
+	return res;
+}
+
+/*
+ * This function may leave a few mappings behind on error, but that's taken
+ * care of by tee_ta_init_user_ta_session() since the entire context is
+ * removed then.
+ */
+TEE_Result ldelf_load_ldelf(struct user_mode_ctx *uctx)
+{
+	TEE_Result res = TEE_SUCCESS;
+	vaddr_t stack_addr = 0;
+	vaddr_t code_addr = 0;
+	vaddr_t rw_addr = 0;
+
+	uctx->is_32bit = is_arm32;
+
+	res = alloc_and_map_ldelf_fobj(uctx, LDELF_STACK_SIZE,
+				       TEE_MATTR_URW | TEE_MATTR_PRW,
+				       &stack_addr);
+	if (res)
+		return res;
+	uctx->ldelf_stack_ptr = stack_addr + LDELF_STACK_SIZE;
+
+	res = alloc_and_map_ldelf_fobj(uctx, ldelf_code_size, TEE_MATTR_PRW,
+				       &code_addr);
+	if (res)
+		return res;
+	uctx->entry_func = code_addr + ldelf_entry;
+
+	rw_addr = ROUNDUP(code_addr + ldelf_code_size, SMALL_PAGE_SIZE);
+	res = alloc_and_map_ldelf_fobj(uctx, ldelf_data_size,
+				       TEE_MATTR_URW | TEE_MATTR_PRW, &rw_addr);
+	if (res)
+		return res;
+
+	vm_set_ctx(uctx->ts_ctx);
+
+	memcpy((void *)code_addr, ldelf_data, ldelf_code_size);
+	memcpy((void *)rw_addr, ldelf_data + ldelf_code_size, ldelf_data_size);
+
+	res = vm_set_prot(uctx, code_addr,
+			  ROUNDUP(ldelf_code_size, SMALL_PAGE_SIZE),
+			  TEE_MATTR_URX);
+	if (res)
+		return res;
+
+	DMSG("ldelf load address %#"PRIxVA, code_addr);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ldelf_init_with_ldelf(struct ts_session *sess __maybe_unused,
+				 struct user_mode_ctx *uctx)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct ldelf_arg *arg = NULL;
+	uint32_t panic_code = 0;
+	uint32_t panicked = 0;
+	uaddr_t usr_stack = 0;
+
+	usr_stack = uctx->ldelf_stack_ptr;
+	usr_stack -= ROUNDUP(sizeof(*arg), STACK_ALIGNMENT);
+	arg = (struct ldelf_arg *)usr_stack;
+	memset(arg, 0, sizeof(*arg));
+	arg->uuid = uctx->ts_ctx->uuid;
+
+	res = thread_enter_user_mode((vaddr_t)arg, 0, 0, 0,
+				     usr_stack, uctx->entry_func,
+				     is_arm32, &panicked, &panic_code);
+
+	thread_user_clear_vfp(&uctx->vfp);
```

Comment:
shippable CI build failure => needs `#if defined(CFG_WITH_VFP)` per @`vfp` field definition.

Reply:
Aha, that's why we had a local `user_clear_vfp()`.

Reply:
Perhaps `thread_user_clear_vfp()` should be passed `uctx` instead?

Reply:
Yes, it could be worth an attempt.

Reply:
I added a new commit (daded14) implementing this. If it looks good, I'll rebase all other commits on top of this.

Reply:
Shippable CI failed with merge conflict error, I'll rebase my whole branch on master.